coprocessor/fhevm-engine/transaction-sender/src/ops/delegate_user_decrypt.rs

```diff
@@ -0,0 +1,705 @@
+            contract_address: delegation.contract_address,
+            delegation_counter: delegation.delegation_counter as u64,
+            old_expiration_date: delegation.old_expiration_date as u64,
+            new_expiration_date: delegation.new_expiration_date as u64,
+            host_chain_id: delegation.host_chain_id as u64,
+            block_hash: delegation.block_hash,
+            block_number: delegation.block_number as u64,
+            transaction_id: delegation.transaction_id,
+            gateway_nb_attempts: delegation.gateway_nb_attempts as u64,
+        };
+        delegations.push(delegation);
+    }
+    Ok(delegations) // delegations)
+}
+
+pub async fn update_error_delegation(
+    tx: &mut DbTransaction<'_>,
+    delegation: &DelegationRow,
+    error: &str,
+) -> () {
+    // update on_gateway
+    error!(%error, ?delegation, "Updating error delegation");
+    let Ok(res) = sqlx::query!(
+        r#"
+        UPDATE delegate_user_decrypt
+        SET gateway_nb_attempts = gateway_nb_attempts + 1,
+            gateway_last_error = $1
+        WHERE key = $2
+        "#,
+        error,
+        delegation.key,
+    )
+    .execute(tx.deref_mut())
+    .await
+    else {
+        error!(error, ?delegation, "Cannot update error delegation");
+        return;
+    };
+    if res.rows_affected() == 0 {
+        error!(
+            error,
+            ?delegation,
+            "No rows updated when updating error delegation"
+        );
+    }
+}
+
+pub async fn update_transmitted_delegation(
+    tx: &mut DbTransaction<'_>,
+    delegation: &DelegationRow,
+) -> () {
+    // update on_gateway
+    let Ok(res) = sqlx::query!(
+        r#"
+        UPDATE delegate_user_decrypt
+        SET on_gateway = true
+        WHERE key = $1
+        "#,
+        delegation.key,
+    )
+    .execute(tx.deref_mut())
+    .await
+    else {
+        error!(?delegation, "Cannot update transmitted delegation");
+        return;
+    };
+    if res.rows_affected() == 0 {
+        error!(
+            ?delegation,
+            "No rows updated when updating transmitted delegation"
+        );
+    }
+}
+
+pub async fn update_useless_delegations(
+    tx: &mut DbTransaction<'_>,
+    clean_before_block: u64,
+    reorg_out_blocks: &[Vec<u8>],
+) -> Result<()> {
+    // update reorg out
+    let reorg_out = sqlx::query!(
+        r#"
+        UPDATE delegate_user_decrypt
+        SET reorg_out = true
+        WHERE block_hash IN (SELECT unnest($1::bytea[]))
+        "#,
+        reorg_out_blocks,
+    )
+    .execute(tx.deref_mut())
+    .await?;
+    if reorg_out.rows_affected() == 0 {
+        error!(
+            reorg_out_blocks = ?reorg_out_blocks,
+            "No rows updated when updating reorg out delegation"
+        );
+    }
+    // clean table past blocks except for errors
+    let cleaned = sqlx::query!(
+        r#"
+        DELETE FROM delegate_user_decrypt
+        WHERE block_number < $1
+        AND gateway_nb_attempts = 0
```

Comment:
any delegate_user_decrypt that had a retry will not be cleaned even if it succeed?